home_audit.ts

```diff
@@ -0,0 +1,691 @@
+ajv.addKeyword('error_msg');
+
+const nullOrUndefined = [null, undefined];
+
+const mandatoryMessage = "Missing value for mandatory field";
+
+module.exports = getNestedValidationMessages;
+
+/**
+ * Perform the HES validation for the nested JSON format. Uses the JSON Schema for initial required field and field
+ * limit validations (e.g. enums, within bounds, etc) and then performs secondary cross validation across the building
+ * as a whole (e.g. the roof area is large enough to cover the floor area)
+ * @param {object} homeValues - JSON object which follows the HES Nested JSON Schema
+ * @return {{blocker: [{string, [string]}], error: [{string, [string]}], warning: [{string, [string]}]}} - Error messages
+ * for the homeValues object, grouped by severity (blocker, error, warning). Messages in each severity group are grouped by
+ * path in the JSON Schema to the error, and can contain multiple errors for a single item in the JSON Schema.
+ */
+let _errorMessages = {};
+function getNestedValidationMessages (homeValues) {
+    _errorMessages = {}
+
+    if(!ajv.validate(HesJsonSchema, homeValues)){
+        ajv.errors.forEach((error) => {
+            const {instancePath, params} = error;
+            const errorPath = params.missingProperty ? `${instancePath}/${params.missingProperty}` : instancePath;
+            addErrorMessage(errorPath, convertAJVError(error))
+        });
+    }
+    getCrossValidationMessages(homeValues);
+    return _errorMessages
+}
+
+
+/**
+ * Convert the AJV error into an intelligible error message that the HES system knows how to display
+ * @param {object} errorObj
+ * @return {string|undefined}
+ */
+function convertAJVError(errorObj) {
+    const {keyword, message, schemaPath} = errorObj;
+    const keyArr = schemaPath.split('/');
+    keyArr.shift(); // remove '#'
+
+    // If it's a keyword that's too deep we should pop it to get the right
+    // level for the error message
+    if(['required', 'const'].includes(keyword)) {
+        keyArr.pop();
+    }
+
+    const error_leaf = keyArr.reduce((acc, key) => {
+        return acc[key]
+    }, HesJsonSchema);
+
+    if(error_leaf.error_msg) {
+        return error_leaf.error_msg;
+    }
+
+    switch(keyword) {
+        case 'required':
+            return mandatoryMessage;
+        case 'enum':
+            return `${message}: '${error_leaf.join('\', \'')}'`;
+        case 'if':
+        case 'not':
+        case 'additionalProperties':
+            // Some keywords indicate errors that we don't want to include in our output
+            return undefined;
+        default:
+            return message;
+    }
+}
+
+/**
+ * Get Cross Validation messages for the building to check for other errors.
+ * @param {object} homeValues
+ */
+function getCrossValidationMessages (homeValues) {
+    getAboutObjectCrossValidationMessages(homeValues)
+    getZoneCrossValidationMessages(homeValues.zone, homeValues.about);
+    getSystemCrossValidation(homeValues.systems)
```

Comment:
Missing semi-colons :(
I haven't decided how particular I should be about this with TS, but it still bothers me

Reply:
According to the hivemind, the same edge-cases that make leaving out semicolons in JS also apply to TS, so we should definitely have them.